batched/system.go

```diff
@@ -0,0 +1,358 @@
+
+	totalProvides, lastReprovideBatchSize     int
+	avgProvideDuration, lastReprovideDuration time.Duration
+}
+
+var _ provider.System = (*BatchProvidingSystem)(nil)
+
+type provideMany interface {
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
+	Ready() bool
+}
+
+// Option defines the functional option type that can be used to configure
+// BatchProvidingSystem instances
+type Option func(system *BatchProvidingSystem) error
+
+var lastReprovideKey = datastore.NewKey("/provider/reprovide/lastreprovide")
+
+func New(provider provideMany, q *queue.Queue, opts ...Option) (*BatchProvidingSystem, error) {
+	s := &BatchProvidingSystem{
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		dynamicCh:         make(chan cid.Cid, 1),
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	// This is after the options processing so we do not have to worry about leaking a context if there is an
+	// initialization error processing the options
+	ctx, cancel := context.WithCancel(context.Background())
+	s.ctx = ctx
+	s.close = cancel
+
+	return s, nil
+}
+
+func Datastore(batching datastore.Batching) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.ds = batching
+		return nil
+	}
+}
+
+func ReproviderInterval(duration time.Duration) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.reprovideInterval = duration
+		return nil
+	}
+}
+
+func KeyProvider(fn simple.KeyChanFunc) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.keyProvider = fn
+		return nil
+	}
+}
+
+func (s *BatchProvidingSystem) Run() {
+	const pauseDetectionThreshold = time.Millisecond * 500
+	const maxCollectionDuration = time.Minute * 10
+	const maxTimeWaitingForProvides = time.Hour
+
+	provCh := s.q.Dequeue()
+
+	go func() {
+		s.closewg.Add(1)
+		defer s.closewg.Done()
+
+		m := make(map[cid.Cid]struct{})
+
+		maxCollectionDurationTimer := time.NewTimer(maxCollectionDuration)
+		pauseDetectTimer := time.NewTimer(maxTimeWaitingForProvides)
+		defer maxCollectionDurationTimer.Stop()
+		defer pauseDetectTimer.Stop()
+
+		for {
+			// Reset timers
+			if !pauseDetectTimer.Stop() {
+				<-pauseDetectTimer.C
+			}
+			pauseDetectTimer.Reset(maxTimeWaitingForProvides)
+
+			if !maxCollectionDurationTimer.Stop() {
+				<-maxCollectionDurationTimer.C
+			}
+			maxCollectionDurationTimer.Reset(maxCollectionDuration)
+
+			performedReprovide := false
+		loop:
+			for {
+				select {
+				case c := <-provCh:
+					m[c] = struct{}{}
+					pauseDetectTimer.Reset(pauseDetectionThreshold)
+					continue
+				default:
+				}
+
+				select {
+				case c := <-provCh:
+					m[c] = struct{}{}
+					pauseDetectTimer.Reset(pauseDetectionThreshold)
+				case c := <-s.dynamicCh:
+					m[c] = struct{}{}
+					pauseDetectTimer.Reset(pauseDetectionThreshold)
```

Comment:
Stop the pause detection timer before restarting it (draining the channel if necessary) if you want to be absolutely sure everything is working correctly (i'd lift this into a helper function).